x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/optimizer/OptimizerTests.java

```diff
@@ -1241,6 +1267,173 @@ public void testRangesOverlappingConjunctionNoLowerBoundary() {
+        Expression exp = rule.rule(and);
+        assertEquals(Range.class, exp.getClass());
+        Range r = (Range) exp;
+        assertEquals(TWO, r.lower());
+        assertFalse(r.includeLower());
+        assertEquals(THREE, r.upper());
+        assertFalse(r.includeUpper());
+    }
+
+    // a != 2 AND 1 < a < 3
+    public void testCombineBinaryComparisonsConjunction_Neq2AndRangeGt1Lt3() {
+        FieldAttribute fa = getFieldAttribute();
+
+        NotEquals neq = new NotEquals(EMPTY, fa, TWO);
+        Range range = new Range(EMPTY, fa, ONE, false, THREE, false);
+        And and = new And(EMPTY, neq, range);
+
+        CombineBinaryComparisons rule = new CombineBinaryComparisons();
+        Expression exp = rule.rule(and);
+        assertEquals(And.class, exp.getClass()); // can't optimize
+    }
+
+    // a != 2 AND a > 3 -> a > 3
+    public void testCombineBinaryComparisonsConjunction_Neq2AndGt3() {
+        FieldAttribute fa = getFieldAttribute();
+
+        NotEquals neq = new NotEquals(EMPTY, fa, TWO);
+        GreaterThan gt = new GreaterThan(EMPTY, fa, THREE);
+        And and = new And(EMPTY, neq, gt);
+
+        CombineBinaryComparisons rule = new CombineBinaryComparisons();
+        Expression exp = rule.rule(and);
+        assertEquals(gt, exp);
+    }
+
+    // a != 2 AND a >= 2 -> a > 2
+    public void testCombineBinaryComparisonsConjunction_Neq2AndGte2() {
+        FieldAttribute fa = getFieldAttribute();
+
+        NotEquals neq = new NotEquals(EMPTY, fa, TWO);
+        GreaterThanOrEqual gte = new GreaterThanOrEqual(EMPTY, fa, TWO);
+        And and = new And(EMPTY, neq, gte);
+
+        CombineBinaryComparisons rule = new CombineBinaryComparisons();
+        Expression exp = rule.rule(and);
+        assertEquals(GreaterThan.class, exp.getClass());
+        GreaterThan gt = (GreaterThan) exp;
+        assertEquals(TWO, gt.right());
+    }
+
+    // a != 2 AND a >= 1 -> nop
+    public void testCombineBinaryComparisonsConjunction_Neq2AndGte1() {
+        FieldAttribute fa = getFieldAttribute();
+
+        NotEquals neq = new NotEquals(EMPTY, fa, TWO);
+        GreaterThanOrEqual gte = new GreaterThanOrEqual(EMPTY, fa, ONE);
+        And and = new And(EMPTY, neq, gte);
+
+        CombineBinaryComparisons rule = new CombineBinaryComparisons();
+        Expression exp = rule.rule(and);
+        assertEquals(And.class, exp.getClass()); // can't optimize
+    }
+
+    // a != 2 AND a <= 3 -> nop
+    public void testCombineBinaryComparisonsConjunction_Neq2AndLte3() {
+        FieldAttribute fa = getFieldAttribute();
+
+        NotEquals neq = new NotEquals(EMPTY, fa, TWO);
+        LessThanOrEqual lte = new LessThanOrEqual(EMPTY, fa, THREE);
+        And and = new And(EMPTY, neq, lte);
+
+        CombineBinaryComparisons rule = new CombineBinaryComparisons();
+        Expression exp = rule.rule(and);
+        assertEquals(And.class, exp.getClass()); // can't optimize
```

Comment:
Wouldn't here be more correct to also assert that the expression resulted from applying the rule is the same And as before? Meaning, is it enough to assume the type of the expression is `And`?

Reply:
That'd be a stricter check than necessary: if the evaluation would build another And instance, the test would still be correct.
Otoh, given the implementation, it's safe to assume that it should be the same And instance, so I've taken on your suggestion.